### PR TITLE
[core][autoscaler][v2] Autoscaler monitor

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR makes Autoscaler V2 works e2e with many of the existing tests on FakeNodeProvider and local cluster mode.

1. adds a V2 Monitor class for running the autoscaler in a loop. This allows e2e testing of the autoscaler with fake node providers. It's compatible with V1's monitor. 
2. add v2 testing in all places where autoscaling is involved (with AutoscalingCluster). Some tests were modified in v2 which I commented on the reasons.
3. clean up logging in v2. 
4. Add ReadOnlyCloudProvider and ReadOnlyConfigReader for v2 when users run ray cluster locally. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/v2/monitor.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+                self._report_autoscaling_state(self.gcs_client, autoscaling_state)
+            else:
+                logger.warning("No autoscaling state to report.")
+
+            # Wait for a autoscaler update interval before processing the next
+            # round of messages.
+            time.sleep(AUTOSCALER_UPDATE_INTERVAL_S)
+
+    def run(self):
+        # FIXME:
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/monitor.py:171]: this depends on the fix in https://github.com/ray-project/ray/pull/43133

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 2]
```diff
@@ -50,117 +48,6 @@
 logger = logging.getLogger(__name__)
 
 
-class IInstanceUpdater(ABC):
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:53]: from previous merge (dead code)

File: python/ray/autoscaler/v2/instance_manager/subscribers/ray_stopper.py [hunk 3]
```diff
@@ -104,17 +104,22 @@ def _drain_ray_node(
...
-            reason=reason,
-            reason_message=reason_str,
-            # TODO: we could probably add a deadline here that's derived
-            # from the stuck instance reconcilation configs.
-            deadline_timestamp_ms=0,
-        )
-        logger.info(f"Draining ray on {ray_node_id}(success={accepted}): {reason_str}")
-        if not accepted:
-            error_queue.put_nowait(RayStopError(im_instance_id=instance_id))
+        try:
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/subscribers/ray_stopper.py:107]: we should also report stop errors if exceptions happen.

File: python/ray/tests/test_chaos.py [hunk 4]
```diff
@@ -305,45 +305,56 @@ def check():
...
+                    "min_workers": 2,
+                    "max_workers": 2,
                 },
-                "node_config": {},
-                "min_workers": 2,
-                "max_workers": 2,
             },
-        },
-    )
-    cluster.start()
```
Comment (rickyyx) [inline python/ray/tests/test_chaos.py:323]: we should always clean up.

**Comment by rickyyx** [comment]
hmm this PR is getting larger now - let me know if you want me to break it down cc @jjyao

File: dashboard/modules/event/tests/test_event.py [hunk 5]
```diff
@@ -294,7 +294,8 @@ async def _check_events(expect_events, read_events, timeout=10):
         assert len(os.listdir(temp_dir)) > 1, "Event log should have rollovers."
 
 
-def test_autoscaler_cluster_events(shutdown_only):
+@pytest.mark.parametrize("autoscaler_v2", [False, True], ids=["v1", "v2"])
```
Comment (jjyao) [inline dashboard/modules/event/tests/test_event.py:297]: what's `ids`?

File: python/ray/autoscaler/v2/autoscaler.py [hunk 6]
```diff
@@ -82,15 +89,22 @@ def _init_cloud_provider(
...
+            node_provider_v1 = _get_node_provider(
+                provider_config,
+                config.get_config("cluster_name"),
+            )
 
-        self._cloud_provider = NodeProviderAdapter(
-            v1_provider=node_provider_v1,
-            config_reader=config_reader,
-        )
+            self._cloud_provider = NodeProviderAdapter(
```
Comment (jjyao) [inline python/ray/autoscaler/v2/autoscaler.py:104]: cloud_instance_provider?

File: python/ray/autoscaler/v2/utils.py [hunk 7]
```diff
@@ -823,3 +823,16 @@ def is_autoscaler_v2() -> bool:
...
+    """
+    Check if the node is a head node from the node state.
+
+    Args:
+        node_state: the node state
+    Returns:
+        is_head: True if the node is a head node, False otherwise.
+    """
+    # TODO: we should include this bit of information in the future.
+    return "node:__internal_head__" in dict(node_state.total_resources)
```
Comment (jjyao) [inline python/ray/autoscaler/v2/utils.py:838]: how does v1 handle it?

File: python/ray/autoscaler/v2/instance_manager/config.py [hunk 8]
```diff
@@ -125,18 +132,28 @@ class AutoscalingConfig:
...
         """
         self._sync_continuously = False
-        self.update_configs(configs, skip_content_hash)
+        self.update_configs(configs, skip_content_hash, skip_prepare)
 
-    def update_configs(self, configs: Dict[str, Any], skip_content_hash: bool) -> None:
+    def update_configs(
+        self, configs: Dict[str, Any], skip_content_hash: bool, skip_prepare
+    ) -> None:
+        if skip_prepare:
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/config.py:154]: what if we just validate as usual for read-only?

File: python/ray/autoscaler/v2/instance_manager/cloud_providers/read_only/cloud_provider.py [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    def get_non_terminated(self) -> Dict[str, CloudInstance]:
+        cluster_resource_state = get_cluster_resource_state(self._gcs_client)
+        cloud_instances = {}
+        for gcs_node_state in cluster_resource_state.node_states:
+            if gcs_node_state.status == NodeStatus.DEAD:
+                # Skip dead nodes.
+                continue
+
+            # Use node's node id if instance id is not available
+            cloud_instance_id = (
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/cloud_providers/read_only/cloud_provider.py:44]: there is code down below that always use node id as cloud instance id

File: python/ray/autoscaler/v2/instance_manager/reconciler.py [hunk 10]
```diff
@@ -769,12 +656,16 @@ def _handle_ray_status_transition(
...
             else:
-                # This should only happen to a ray node that's not managed by us.
-                logger.warning(
-                    f"Ray node {binary_to_hex(n.node_id)} has no instance id. "
-                    "This only happens to a ray node that's not managed by autoscaler. "
-                    "If not, please file a bug at https://github.com/ray-project/ray"
-                )
+                if autoscaling_config.provider == Provider.READ_ONLY:
+                    # We will use the node id as the cloud instance id for read-only provider.
+                    ray_nodes_by_cloud_instance_id[binary_to_hex(n.node_id)] = n
```
Comment (jjyao) [inline python/ray/autoscaler/v2/instance_manager/reconciler.py:661]: here

File: python/ray/autoscaler/v2/monitor.py [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    def __init__(
+        self,
+        gcs_address: str,
+        config_reader: IConfigReader,
+        log_dir: str = None,
+        monitor_ip: str = None,
+    ):
+        self.gcs_address = gcs_address
+        worker = ray._private.worker.global_worker
+        # TODO: eventually plumb ClusterID through to here
```
Comment (jjyao) [inline python/ray/autoscaler/v2/monitor.py:65]: ?

File: python/ray/autoscaler/v2/monitor.py [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+
+    @staticmethod
+    def _get_session_name(gcs_client: GcsClient) -> Optional[str]:
+        """Obtain the session name from the GCS.
+
+        If the GCS doesn't respond, session name is considered None.
+        In this case, the metrics reported from the monitor won't have
+        the correct session name.
+        """
+        if not _internal_kv_initialized():
```
Comment (jjyao) [inline python/ray/autoscaler/v2/monitor.py:134]: ?

File: python/ray/autoscaler/v2/tests/test_e2e.py [hunk 13]
```diff
@@ -236,7 +242,8 @@ def pg_created():
         cluster.shutdown()
 
 
-def test_placement_group_removal_idle_node():
+@pytest.mark.parametrize("autoscaler_v2", [False, True], ids=["v1", "v2"])
```
Comment (jjyao) [inline python/ray/autoscaler/v2/tests/test_e2e.py:245]: ray_start_autoscaling_cluster fixture

**Review by jjyao — APPROVED** [review]

File: python/ray/autoscaler/v2/monitor.py (hunk 11 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/monitor.py:65]: from cluster id work.

File: python/ray/autoscaler/v2/tests/test_e2e.py (hunk 13 above)
Comment (rickyyx) [reply python/ray/autoscaler/v2/tests/test_e2e.py:245]: https://github.com/ray-project/ray/issues/43723

**Pull request merged**